File: backend/toms_gym/services/bowling_score.py
"""Ten-pin scoring from printed roll symbols. Pure, DB-free."""


class InvalidFrame(ValueError):
    pass


def _pins(symbol, prev):
    s = str(symbol).strip().upper()
    if s == "X":
        return 10
    if s == "/":
        if prev is None:
            raise InvalidFrame("spare with no first ball")
        return 10 - prev
    if s in ("-", "F", "", "O", "_"):
        return 0
    if s.isdigit():
        v = int(s)
        if v > 9:
            raise InvalidFrame(f"pin count {v} out of range")
        return v
    raise InvalidFrame(f"unknown symbol {symbol!r}")


def frame_pins(rolls, frame_index):
    """Symbols for one frame -> pins felled per ball."""
    if not rolls:
        raise InvalidFrame("empty frame")
    tenth = frame_index == 9
    if len(rolls) > (3 if tenth else 2):
        raise InvalidFrame("too many rolls")
    pins = []
    prev = None
    for sym in rolls:
        p = _pins(sym, prev)
        if prev is not None and str(sym).strip().upper() != "/" and prev + p > 10:
            raise InvalidFrame(f"{prev}+{p} exceeds 10")
        pins.append(p)
        # a strike or a spare resets the rack; otherwise the next ball completes it
        prev = None if (p == 10 or prev is not None) else p
    if not tenth:
        if pins[0] == 10 and len(pins) > 1:
            raise InvalidFrame("a strike ends the frame")
        if len(rolls) == 1 and pins[0] != 10:
            raise InvalidFrame("open frame needs two balls")
    else:
        need_third = pins[0] == 10 or (len(pins) >= 2 and pins[0] + pins[1] == 10)
        if need_third and len(pins) < 3:
            raise InvalidFrame("tenth frame bonus ball missing")
        if not need_third and len(pins) == 3:
            raise InvalidFrame("tenth frame has an extra ball")
    return pins


def score_frames(frames):
    """Score up to ten frames. Never raises: bad frames land in ``errors``."""
    frames = list(frames)[:10]
    rolls = []
    errors = []
    for i, f in enumerate(frames):
        try:
            p = frame_pins(f, i)
        except InvalidFrame as e:
            errors.append(f"frame {i + 1}: {e}")
            p = [0, 0]
        rolls.extend(p)
    cumulative = []
    total = 0
    i = 0
    complete = len(frames) == 10
    for fr in range(len(frames)):
        if i >= len(rolls):
            complete = False
            break
        first = rolls[i]
        if fr == 9:
            total += sum(rolls[i:])
            i = len(rolls)
        elif first == 10:
            bonus = rolls[i + 1:i + 3]
            if len(bonus) < 2:
                complete = False
            total += 10 + sum(bonus)
            i += 1
        elif i + 1 < len(rolls) and first + rolls[i + 1] == 10:
            bonus = rolls[i + 2:i + 3]
            if not bonus:
                complete = False
            total += 10 + sum(bonus)
            i += 2
        else:
            total += first + (rolls[i + 1] if i + 1 < len(rolls) else 0)
            i += 2
        cumulative.append(total)
    return {
        "cumulative": cumulative,
        "total": total,
        "valid": not errors,
        "errors": errors,
        "complete": complete,
    }


def validate_night_row(games, scratch, hdcp, total):
    """Check the two printed checksums on a night-results row."""
    present = [g for g in games if g is not None]
    if len(present) != len(games):
        return False, "missing game score"
    if scratch is not None and sum(present) != scratch:
        return False, f"games sum {sum(present)} != scratch {scratch}"
    if scratch is not None and hdcp is not None and total is not None and scratch + hdcp != total:
        return False, f"scratch {scratch} + hdcp {hdcp} != total {total}"
    return True, None


def infer_night_row(games, scratch, hdcp, total):
    """Fill values the checksums determine uniquely. Returns (row, inferred_keys)."""
    games = list(games)
    inferred = []
    missing = [i for i, g in enumerate(games) if g is None]
    if len(missing) == 1 and scratch is not None:
        i = missing[0]
        games[i] = scratch - sum(g for g in games if g is not None)
        inferred.append(f"games[{i}]")
    elif not missing and scratch is None:
        scratch = sum(games)
        inferred.append("scratch")
    if scratch is not None:
        if total is None and hdcp is not None:
            total = scratch + hdcp
            inferred.append("total")
        elif hdcp is None and total is not None:
            hdcp = total - scratch
            inferred.append("hdcp")
    return {"games": games, "scratch": scratch, "hdcp": hdcp, "total": total}, inferred

File: frontend/src/lib/bowlingScore.ts
/**
 * Ten-pin scoring from printed roll symbols.
 * Port of backend `services/bowling_score.py::score_frames` — the review page
 * scores edits live, the confirm route re-scores server-side, and the two must
 * agree symbol for symbol.
 */

export type Roll = string;

export interface ScoredFrames {
  cumulative: number[];
  total: number;
  valid: boolean;
  errors: string[];
  complete: boolean;
}

export class InvalidFrame extends Error {
  constructor(message: string) {
    super(message);
    this.name = "InvalidFrame";
    Object.setPrototypeOf(this, InvalidFrame.prototype);
  }
}

const normalize = (symbol: Roll): string => symbol.trim().toUpperCase();

const sum = (values: number[]): number => values.reduce((a, b) => a + b, 0);

function pinsFor(symbol: Roll, standing: number | null): number {
  const s = normalize(symbol);
  if (s === "X") return 10;
  if (s === "/") {
    if (standing === null) throw new InvalidFrame("spare with no first ball");
    return 10 - standing;
  }
  if (s === "-" || s === "F" || s === "") return 0;
  if (/^\d+$/.test(s)) {
    const v = Number(s);
    if (v > 9) throw new InvalidFrame(`pin count ${v} out of range`);
    return v;
  }
  throw new InvalidFrame(`unknown symbol "${symbol}"`);
}

/** Pins knocked down per ball. Throws `InvalidFrame` on an impossible frame. */
export function framePins(rolls: Roll[], frameIndex: number): number[] {
  if (rolls.length === 0) throw new InvalidFrame("empty frame");
  const tenth = frameIndex === 9;
  if (rolls.length > (tenth ? 3 : 2)) throw new InvalidFrame("too many rolls");

  const pins: number[] = [];
  // Pins felled by the first ball of the rack in play; null = fresh rack.
  let standing: number | null = null;
  for (const symbol of rolls) {
    const p = pinsFor(symbol, standing);
    if (standing !== null) {
      if (standing + p > 10) throw new InvalidFrame(`${standing}+${p} exceeds 10`);
      standing = null;
    } else if (p < 10) {
      standing = p;
    }
    pins.push(p);
  }

  if (!tenth) {
    if (pins.length === 1 && pins[0] !== 10) {
      throw new InvalidFrame("open frame needs two balls");
    }
    if (pins.length === 2 && pins[0] === 10) {
      throw new InvalidFrame("a strike ends the frame");
    }
    return pins;
  }

  const needsThird = pins[0] === 10 || (pins.length >= 2 && pins[0] + pins[1] === 10);
  if (needsThird && pins.length < 3) {
    throw new InvalidFrame("tenth frame bonus ball missing");
  }
  if (!needsThird && pins.length === 3) {
    throw new InvalidFrame("tenth frame has an extra ball");
  }
  if (pins.length === 1) throw new InvalidFrame("tenth frame needs two balls");
  return pins;
}

/**
 * Running score for up to ten frames. Invalid frames are reported in `errors`
 * (and scored as an open zero) rather than thrown, so a half-typed review row
 * still renders. `complete` is false while any bonus ball is still unthrown.
 */
export function scoreFrames(frames: Roll[][]): ScoredFrames {
  const capped = frames.slice(0, 10);
  const rolls: number[] = [];
  const errors: string[] = [];

  capped.forEach((frame, index) => {
    try {
      rolls.push(...framePins(frame, index));
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      errors.push(`frame ${index + 1}: ${message}`);
      rolls.push(0, 0);
    }
  });

  const cumulative: number[] = [];
  let total = 0;
  let i = 0;
  let complete = capped.length === 10;

  for (let frame = 0; frame < capped.length; frame += 1) {
    if (i >= rolls.length) break;
    const first = rolls[i];
    if (frame === 9) {
      total += sum(rolls.slice(i));
      i = rolls.length;
    } else if (first === 10) {
      const bonus = rolls.slice(i + 1, i + 3);
      if (bonus.length < 2) complete = false;
      total += 10 + sum(bonus);
      i += 1;
    } else if (i + 1 < rolls.length && first + rolls[i + 1] === 10) {
      const bonus = rolls.slice(i + 2, i + 3);
      if (bonus.length === 0) complete = false;
      total += 10 + sum(bonus);
      i += 2;
    } else {
      total += first + (i + 1 < rolls.length ? rolls[i + 1] : 0);
      i += 2;
    }
    cumulative.push(total);
  }

  return { cumulative, total, valid: errors.length === 0, errors, complete };
}

/**
 * Split a typed cell ("X", "9/", "XX8", "F-") into roll symbols. Frames 1-9
 * take two balls, the tenth takes three.
 */
export function parseFrameInput(text: string, frameIndex: number): Roll[] {
  const max = frameIndex === 9 ? 3 : 2;
  return text
    .toUpperCase()
    .split("")
    .filter((c) => c !== " ")
    .slice(0, max);
}

/** Render roll symbols back into a single editable cell string. */
export function formatFrameInput(rolls: Roll[] | null | undefined): string {
  return (rolls || []).map(normalize).join("");
}
